### PR TITLE
💜 Abstract persistence layer

### PR DESCRIPTION
**Pull request by skittlz444**
LOTS of file changes sorry. Should still be thoroughly code reviewed before merging, better yet, pull down the pull request branch and test/code review it there.

### CODE REVIEW

**Comment by Breadkenty** [comment]
This is starting to make more sense, is there a reason why there are two ClientApp folders? One with the graphics folder, and another in the API folder which also has the graphics folder in the src.

**Comment by skittlz444** [comment]
That's probably an issue cleaning up the folders on checkout, I think it's because node modules is ignored and therefore not cleaned. Remove the one not in the api folder.

File: Api/Properties/launchSettings.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+      "launchBrowser": true,
+      "environmentVariables": {
+        "ASPNETCORE_ENVIRONMENT": "Development"
+      }
+    },
+    "Smash_Combos": {
+      "commandName": "Project",
+      "launchBrowser": true,
+      "environmentVariables": {
+        "DATABASE_URL": "postgres://postgres:Skittlz0964@localhost:5432/SmashCombos_Database",
```
Comment (Breadkenty) [inline Api/Properties/launchSettings.json:22]: I think I'm having issues with this line where the path of my database being wrong. Correct my if I'm wrong

```
Unhandled exception. Npgsql.PostgresException (0x80004005): 28000: role "postgres" does not exist
   at Npgsql.NpgsqlConnector.<>c__DisplayClass160_0.<<DoReadMessage>g__ReadMessageLong|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Npgsql.NpgsqlConnector.<>c__DisplayClass160_0.<<DoReadMessage>g__ReadMessageLong|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Npgsql.NpgsqlConnector.Open(NpgsqlTimeout timeout, Boolean async, CancellationToken cancellationToken)
   at Npgsql.NpgsqlConnection.<>c__DisplayClass32_0.<<Open>g__OpenLong|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Npgsql.EntityFrameworkCore.PostgreSQL.Storage.Internal.NpgsqlDatabaseCreator.ExistsAsync(CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Migrations.HistoryRepository.ExistsAsync(CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Migrations.HistoryRepository.GetAppliedMigrationsAsync(CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.RelationalDatabaseFacadeExtensions.GetAppliedMigrationsAsync(DatabaseFacade databaseFacade, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.RelationalDatabaseFacadeExtensions.GetPendingMigrationsAsync(DatabaseFacade databaseFacade, CancellationToken cancellationToken)
   at Smash_Combos.Program.Main(String[] args) in /Users/kento/Desktop/SDG/Smash_Combos/Api/Program.cs:line 25
   at Smash_Combos.Program.<Main>(String[] args)
  Exception data:
    Severity: FATAL
    SqlState: 28000
    MessageText: role "postgres" does not exist
    File: miscinit.c
    Line: 607
    Routine: InitializeSessionUserId
```

Comment (skittlz444) [reply Api/Properties/launchSettings.json:22]: Whoops that launch settings was supposed to be gitignored, will remove and ignore

File: Api/Smash_Combos.csproj [hunk 2]
```diff
@@ -28,6 +28,11 @@
     <None Remove="$(SpaRoot)**" />
     <None Include="$(SpaRoot)**" Exclude="$(SpaRoot)node_modules\**" />
   </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\Persistence\Smash_Combos.Persistence.csproj" />
```
[the remark this answers is not in the document]
Comment (skittlz444) [reply Api/Smash_Combos.csproj:32]: Yup, that's the line I missed. That's what I get for coding with wine haha. Thanks for the careful review!

**Pull request merged**